### DISCUSSION TITLE
Welcome to Arduino-BLE-MIDI Discussions!

### DISCUSSION THREAD

**Discussion by lathoub** [lathoub/Arduino-BLE-MIDI#41] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Darce-One** [comment]
Hello, my name is Andreas and I'm using the BLE-MIDI to send data into Max/MSP to create interactive sound boxes.

I've been having an issue with it though recently, where the arduino goes offline before sending any data.
I am using an Arduino Nano IoT 33, and wish to send accelorometer data using the LSM6DS3 module and library. 
Here is the code I've written so far:
[acceleration_test.zip](https://github.com/lathoub/Arduino-BLE-MIDI/files/7735190/acceleration_test.zip)

My guess as to what is happening is that the arduino is overworked and cannot execute the code on time? would anyone know if there is a way around it? This may also not be the issue, I am new to all this and would really appreciate the help.

**Reply by Darce-One** [reply]
I'm happy to say I've sorted it, dumb mistake on my part.

I hope to remember to share what I've done once it's finished. Thanks @lathoub for making this library, it's very useful.

**Reply by agentff6600** [reply]
what was the issue?

**Reply by Darce-One** [reply]
I can't remember what the issue was, there were many little ones haha. two that I remember were 1) not initialising the BLE correctly, and another was the soldering, some circuits were crossing paths and leading to mixed signals crashing the arduino.

**Reply by Darce-One** [reply]
here's the final product!
https://youtu.be/dbJ7OBR2PAY

**Reply by agentff6600** [reply]
super nice!

I tired you sketch that you uploaded here and I seems that I have a similar issue with what you had. Let me explain:

I uploaded your sketch onto a nano 33 IOT.
As soon as I try to connect it via bluetooth in the Audio Midi Setup (on macOS) the LED on the nano blinks once and then turns off again.
I assume that the BLE was immediately disconnected.

I have a similar issue with one of my attempts.

I tried to use the standard MidiBle example and removed the midi send function in the main loop.
After I did this, the nano has the same behaviour like described above.

So long story short: I wonder what you changed in your script to get it running? My hunch is it also has to do with the BLE initialisation.

Any tips? Or do you have the working sketch of your accelerometer project?

**Comment by Spartacus1066** [comment]
It looks like @Darce-One and @agentff6600 have got the library to work BLE MIDI with a Nano 33 iOT, but I cannot. If you guys could share ANY working code, I would much appreciate it.
I *do* have a project working using the library with a Feather ESP32-S3 and the NimBLE bits of this library, but not with the 33 iOT.

Even this SIMPLE code is not compiling.
```#include <Arduino.h>
#include <BLEMIDI_Transport.h>
#include <hardware/BLEMIDI_ArduinoBLE.h>

unsigned long t0 = millis();
bool isConnected = false;

BLEMIDI_CREATE_INSTANCE("TestMIDI", MIDI);

void OnConnected()
{
  isConnected = true;
  digitalWrite(LED_BUILTIN, HIGH);
}
void OnDisconnected()
{
  isConnected = false;
  digitalWrite(LED_BUILTIN, LOW);
}
void OnNoteOn(byte channel, byte note, byte velocity)
{
}
void OnNoteOff(byte channel, byte note, byte velocity)
{
}
void setup()
{

  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, LOW);
  BLEMIDI.setHandleConnected(OnConnected);
  BLEMIDI.setHandleDisconnected(OnDisconnected);
  MIDI.begin();
  MIDI.setHandleNoteOn(OnNoteOn);
  MIDI.setHandleNoteOff(OnNoteOff);
}


void loop()
{
  while (isConnected)
  {
    MIDI.read();
    while (true)
    {
      MIDI.sendNoteOn(65, 127, 1);
      delay(1000);
      MIDI.sendNoteOff(65, 0, 1);
      delay(1000);
    }
  }
}
```
I am getting:
Processing nano_33_iot (platform: atmelsam; board: nano_33_iot; framework: arduino)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/atmelsam/nano_33_iot.html
PLATFORM: Atmel SAM (8.1.0) > NANO 33 IoT
HARDWARE: SAMD21G18A 48MHz, 32KB RAM, 256KB Flash
DEBUG: Current (atmel-ice) External (atmel-ice, blackmagic, jlink)
PACKAGES: 
 - framework-arduino-samd @ 1.8.13 
 - framework-cmsis @ 1.40500.0 (4.5.0) 
 - framework-cmsis-atmel @ 1.2.2 
 - toolchain-gccarmnoneeabi @ 1.70201.0 (7.2.1)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 16 compatible libraries
Scanning dependencies...
Dependency Graph
|-- BLE-MIDI @ 2.2.0
|   |-- MIDI Library @ 5.0.2
|   |-- ArduinoBLE @ 1.3.1
Building in release mode
Compiling .pio\build\nano_33_iot\src\main.cpp.o
In file included from src\main.cpp:3:0:
.pio\libdeps\nano_33_iot\BLE-MIDI\src/hardware/BLEMIDI_ArduinoBLE.h:5:24: error: 'SERVICE_UUID' was not declared in this scope
 BLEService midiService(SERVICE_UUID);
                        ^~~~~~~~~~~~
.pio\libdeps\nano_33_iot\BLE-MIDI\src/hardware/BLEMIDI_ArduinoBLE.h:5:24: note: suggested alternative:
In file included from src\main.cpp:2:0:
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h:21:26: note:   'bleMidi::SERVICE_UUID'
 static const char *const SERVICE_UUID        = "03b80e5a-ede8-4b33-a751-6ce34ec4c700";
                          ^~~~~~~~~~~~
In file included from src\main.cpp:3:0:
.pio\libdeps\nano_33_iot\BLE-MIDI\src/hardware/BLEMIDI_ArduinoBLE.h:6:34: error: 'CHARACTERISTIC_UUID' was not declared in this scope
 BLEStringCharacteristic midiChar(CHARACTERISTIC_UUID,  // standard 16-bit characteristic UUID
                                  ^~~~~~~~~~~~~~~~~~~
.pio\libdeps\nano_33_iot\BLE-MIDI\src/hardware/BLEMIDI_ArduinoBLE.h:6:34: note: suggested alternative:
In file included from src\main.cpp:2:0:
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h:22:26: note:   'bleMidi::CHARACTERISTIC_UUID'
 static const char *const CHARACTERISTIC_UUID = "7772e5db-3868-4112-a1a9-f2669d106bf3";
                          ^~~~~~~~~~~~~~~~~~~
In file included from src\main.cpp:2:0:
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h: In instantiation of 'void bleMidi::BLEMIDI_Transport<T, _Settings>::receive(byte*, size_t) [with T = bleMidi::BLEMIDI_ArduinoBLE; _Settings = bleMidi::DefaultSettings; byte = unsigned char; size_t = unsigned int]':
.pio\libdeps\nano_33_iot\BLE-MIDI\src/hardware/BLEMIDI_ArduinoBLE.h:107:65:   required from here
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h:297:60: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             while ((buffer[rPtr + 1] < MIDI_TYPE) && (rPtr < (length - 1)))
                                                      ~~~~~~^~~~~~~~~~~~~~~
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h:391:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             if (++rPtr >= length)
                 ~~~~~~~^~~~~~~~~
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h:408:22: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             if (lPtr >= length)
                 ~~~~~^~~~~~~~~
.pio\libdeps\nano_33_iot\BLE-MIDI\src/BLEMIDI_Transport.h:260:14: warning: unused variable 'timestampHigh' [-Wunused-variable]
         auto timestampHigh = 0x3f & headerByte;
              ^~~~~~~~~~~~~
*** [.pio\build\nano_33_iot\src\main.cpp.o] Error 1
=============================================================================================== [FAILED] Took 1.45 seconds

**Comment by lathoub** [comment]
Use git [branch for Arduino-Nano-BLE-33](https://github.com/lathoub/Arduino-BLE-MIDI/tree/Arduino-Nano-BLE-33) - and let me know if it all work alright.
